### DISCUSSION TITLE
Upgrade para a versão 3.0.17

### DISCUSSION THREAD

**Discussion by rssprog** [viniciussanchez/RESTRequest4Delphi#85] [General]
Caros amigos, boa tarde.

Após a atualização para a versão 3.0.17, meus sistemas pararam de funcionar.
O motivo é o seguinte:
Na chamada da IRequest.Get (ou Put, etc), quando o servidor estava off-line o retorno do meu IResponse vinha preenchido, ou seja, not null (StatusCode=0).
Agora, se o servidor estiver off-line retorna null para o IResponse.
Como eu fazia validações com IResponse.StatusCode, estou recebendo erros em algumas rotinas onde esse tratamento era feito.
Alguém mais passando pelo mesmo problema?

**Comment by viniciussanchez** [comment]
@rssprog até o momento não tive outras reclamações sobre isso... qual versão você usava? Vou fazer um teste aqui tentando simular o que você falou e posto mais informações... mas a princípio basta você continuar usando a versão anterior ate que isso seja verificado

**Comment by viniciussanchez** [comment]
@rssprog deve ter algo de errado... rsrsrs

Fiz a requisição com o servidor off line:
![image](https://user-images.githubusercontent.com/16382981/156437780-e8ed5deb-803a-425d-999f-92fe29896cb4.png)

O response veio criado com o status code = 0
![image](https://user-images.githubusercontent.com/16382981/156437867-c71311aa-082e-4d69-9bbe-c90261169fed.png)

Usei o projeto "client" (delphi) que está na pasta samples do projeto para fazer o teste.

**Reply by rssprog** [reply]
Mestre Vinícius, seguinte.
Estou utilizando o Delphi Tokyo.
Vou fazer um teste isolado das minhas aplicações.

**Reply by viniciussanchez** [reply]
Hum... só se for alguma coisa relacionada a versão de Delphi. Eu estou usando aqui o Delphi 10.3.3

**Reply by rssprog** [reply]
![image](https://user-images.githubusercontent.com/24495822/156440255-7ac1f0c3-7500-4ace-84e4-3708266abbde.png)

Vinicius, me perdoe por tomar o seu tempo.
De fato, como mostra a imagem, isoladamente funcionou!
Aí eu pensei... espere, eu estava usando FMX.
Então tb fiz com o Firemonkey e também passou.
Vou ter que descobrir o que está acontecendo, pois comprovadamente o problema está nas minhas aplicações.

Não herdei e não fiz qualquer alteração nos seus fontes.
Vou perder a noite analisando.

Obrigado, parabéns e forte abraço.

**Reply by viniciussanchez** [reply]
blz... vlw

**Reply by rssprog** [reply]
Vinicius, descobri o problema.
Só que não entendi o motivo de não funcionar mais.

Criei algumas funções para ajudar no rápido desenvolvimento das minhas aplicações.
Assim, para criar o IRequest eu criei uma function com a seguinte assinatura:

function CreateRequest(const AResource: string; ADataSet: TDataSet=nil): IRequest;

A função faz o preenchimento da BaseURL(com base em uma variável, que pode ser produção ou homologação) e do BasicAuthentication, também do DataSetAdapter, que é opcional.

Depois, para cada método, eu criei uma função para validar o request, com a seguinte assinatura:

function RequestGet(var ARequest: IRequest; var AResponse: IResponse; var AErrorMsg: string): Boolean;

Essa função faz todo o tratamento da chamada e sempre retorna True para o StatusCode>=200 e <=399.
Dentro fr bloco try para tratar justamente erros como o servidor off-line.

Dentro do escopo da função, AResponse tem valor, mas após o retorno, a variável AResponse volta como null, o que não acontecia antes.
E me enganei com a versão do Delphi, estou utilizando a Rio (10.3).

Para resolver o problema criei outra função com overload da primeira, com a diferença do AStatusCode como parâmetro:

function RequestGet(var ARequest: IRequest; var AResponse: IResponse; var AStatusCode: integer; var AErrorMsg: string): Boolean;

Corri os códigos onde eu usava o Response.StatusCode  do retorno e troquei para a nova function, resolvendo o problema.

Forte abraço.

**Reply by viniciussanchez** [reply]
Entendi... estranho então ter parado de funcionar assim...
